multi/imxrt-multi/i2c.c

```diff
@@ -1,84 +1,656 @@
+#define STATUS_END              (1u << 8)
+#define STATUS_RXFIFO_WM        (1u << 1)
+#define STATUS_TXFIFO_WM        (1u << 0)
+/* STATUS_STOP is not an error, but we check for it in case of a programming mistake */
+#define STATUS_ERRORS (STATUS_PIN_LOW_TIMEOUT | STATUS_FIFO_ERROR | STATUS_ARBITRATION_LOST | STATUS_NACK | STATUS_STOP)
+
+#define RET_STOP 1
+
+
+/* clang-format off */
+
+enum { veridr = 0, paramr, mcr = 4, msr, mier, mder, mcfgr0, mcfgr1, mcfgr2, mcfgr3, mdmr = 16,
 	mccr0 = 18, mccr1 = 20, mfcr = 22, mfsr, mtdr, mrdr = 28, scr = 68, ssr, sier, sder,
-	scfgr1 = 73, scfgr2, samr = 80, sasr = 84, star, stdr= 88, srdr= 92 };
+	scfgr1 = 73, scfgr2, samr = 80, sasr = 84, star, stdr = 88, srdr = 92 };
+
+enum { cmd_txdata = 0, cmd_rxdata = (1 << 8), cmd_stop = (2 << 8), cmd_start = (4 << 8) };
+
+enum { i2c_stateBusy = 0, i2c_stateReady = 1 };
+
+/* clang-format on */
+
+
+struct periph_info {
+	volatile uint32_t *base;
+
+	handle_t mutex;
+	handle_t irqMutex;
+	handle_t irqCond;
+	handle_t irqHandle;
+	int irqNo;
+
+	volatile int state;
+} i2c_common[I2C_CNT];
+
+
+static int i2c_irqRoutine(unsigned int n, void *arg)
+{
+	TRACE_IRQ();
+	struct periph_info *periph = arg;
+
+	*(periph->base + mier) = 0;
+	periph->state = i2c_stateReady;
+
+	return 0;
+}
+
+
+static inline bool i2c_getTimings(uint8_t speed, uint32_t *mccr0Val, uint32_t *prescalerVal, uint32_t *mcfgr2Val)
+{
+	if ((speed < i2c_speed_slow) || (speed > i2c_speed_fast_plus)) {
+		return false;
+	}
+
+#ifdef __CPU_IMXRT117X
+	/* DATAVD, SETHOLD, CLKHI, CLKLO */
+	static const uint32_t clockConfig[] = {
+		[i2c_speed_slow] = 0x0f1d353eu,
+		[i2c_speed_fast] = 0x03070b11u,
+		[i2c_speed_fast_plus] = 0x04030406u,
+	};
+
+	/* Recommended minimum BUSIDLE values according to documentation */
+	static const uint8_t busIdle[] = {
+		[i2c_speed_slow] = (0x1d + 0x3e + 2) * 2,
+		[i2c_speed_fast] = (0x07 + 0x11 + 2) * 2,
+		[i2c_speed_fast_plus] = (0x03 + 0x06 + 2) * 2,
+	};
+
+	*prescalerVal = 1;
+	const uint32_t filt = 1u;
+#else
+	static const uint32_t clockConfig[] = {
+		[i2c_speed_slow] = 0x08111f28u,
+		[i2c_speed_fast] = 0x08111f28u,
+		[i2c_speed_fast_plus] = 0x01070b0fu,
+	};
+
+	static const uint8_t busIdle[] = {
+		[i2c_speed_slow] = (0x11 + 0x28 + 2) * 2,
+		[i2c_speed_fast] = (0x11 + 0x28 + 2) * 2,
+		[i2c_speed_fast_plus] = (0x07 + 0x0f + 2) * 2,
+	};
+
+	*prescalerVal = (speed == i2c_speed_slow) ? 3u : 1u;
+	const uint32_t filt = 2u;
+#endif
+
+	*mccr0Val = clockConfig[speed];
+	*mcfgr2Val = (filt << 24u) | (filt << 16u) | (uint32_t)busIdle[speed];
+	return true;
+}
+
+
+static int i2c_setupMaster(int pos, uint8_t speed, uint8_t pushpull)
+{
+	uint32_t mccr0Val, prescalerVal, mcfgr2Val;
+	if (!i2c_getTimings(speed, &mccr0Val, &prescalerVal, &mcfgr2Val)) {
+		return -EINVAL;
+	}
+
+	mutexLock(i2c_common[pos].mutex);
+
+	TRACE();
+	volatile uint32_t *base = i2c_common[pos].base;
+
+	/* Reset before configuring */
+	*(base + mcr) = 1u << 1u;
+	*(base + mcr) = 0u;
```

Comment:
Maybe a memory barrier before and after this write?

Reply:
Done